src/java.base/share/classes/sun/security/util/math/IntegerModuloP.java

```diff
@@ -208,5 +211,206 @@ default ImmutableIntegerModuloP pow(BigInteger b) {
+                // is pre-computed to speed up the computation.
+
+                // calculate and cache imp ^ (2^2 - 1) - imp ^ (2^4 - 1)
+                IntegerModuloP[] w = new IntegerModuloP[4];
+                w[0] = imp.fixed();
+                MutableIntegerModuloP t = imp.mutable();
+                for (int i = 1; i < 4; i++) {
+                    t.setSquare();
+                    t.setProduct(imp);
+                    w[i] = t.fixed();
+                }
+
+                // calculate imp ^ (2^32 - 1)
+                MutableIntegerModuloP d = null;
+                for (int i = 4; i < 32; i++) {
+                    t.setSquare();
+                    switch (i) {
+                        case 7 -> {
+                            t.setProduct(w[3]);
+                            d = t.mutable();   // 7: imp ^ (2^8 - 1)
+                        }
+                        case 15 -> {
+                            t.setProduct(d);
+                            d = t.mutable();   // 15: imp ^ (2^16 - 1)
+                        }
+                        case 31 -> {
+                            t.setProduct(d); // 31: imp ^ (2^32 - 1)
+                        }
+                    }
+                }
+
+                // Here we have:
+                //    w[i] = imp ^ (2 ^ ( i + 1) - 1), i = {0, 1, 2, 3}
+                //       t = imp ^ (2^32 - 1)
+                //
+                // calculate for bit 32-128, for contiguous 32-bit set.
+                d = t.mutable();
+                for (int i = 32; i < 128; i++) {
+                    d.setSquare();
+                    if (i == 95 || i == 127) {
+                        d.setProduct(t);
+                    }
+                }
+
+                // Calculate for bit 128-255, for individual 1-bit set.
+                for (int k = -1, i = 127; i >= 0; i--) {
+                    if (b.testBit(i)) {
+                        if (k == w.length - 2) {
+                            // calculate the current & reserved bits
+                            d.setSquare();
+                            d.setProduct(w[w.length - 1]);
+                            k = -1;
+                        } else {
+                            k++;
+                            d.setSquare();
+                        }
+                    } else {    // calculate the reserved bits
+                        if (k >= 0) {
+                            // add back the reserved bits
+                            d.setProduct(w[k]);
+                            k = -1;
+                        }
+                        d.setSquare();
+                    }
+                }
+
+                // Note that the last 4 bits was handled in the for-lopp above
+                // as it hapeens to be 4. For bit set other than 4 bits, for
+                // example, 3 bits set (0x8), the value should be added back.
+                // d.setProduct(w[2]);
```

Comment:
I think you can remove this comment, or at least fix your typos: "for-lopp" -> "for loop",  "happens" -> "happens",  "(0x8)" -> "(0x7)". 
You can say something like.: ' "if(k != -1) d.setProduct(w[k]);" is not necessary here as k is -1 for this exponent'

Reply:
Thank you for the suggestion.  I would like to remove this comment as it looks more clear to me.